### PR TITLE
Fix #12379: Fixed Column Type Validation Logic and added tests for th…

### LINKED ISSUE #12379: [Bug] Unhelpful bug from agate when column type specified for non-existent column
### Housekeeping

- [x] I am a maintainer of dbt-core

### Short description

For seeds you can define column types via something like
```yml
seeds:
  - name: my_seed
    config:
      column_types:
        my_column_a: varchar(2)
        my_column_b: varchar(32)
```

You don't have to specify column types, if you don't, we try to infer what type the column is. Sometimes this inference doesn't think it's the type you know it is, which is why we allow for specifying the type.

A problem arises though when you specify a column type for a column that _doesn't exist_. In our example above let's say that `my_column_b` doesn't exist. Well when we dbt goes to determine column types a warning will emitted from `agate` (which we use for inferring/setting types) like the following

> RuntimeWarning: "my_column_b" does not match the name of any column in this table.

The warning doesn't tell you _which_ seed has the problem. If you only have a few seeds and none have a `my_column_b` solving the problem is quite trivial. Run a search on your YAML files for seeds + your dbt_project.yaml and remove any references to `my_column_b`. But what if you have a lot of seed and some or most have a `my_column_b`? **Well then it becomes a very annoying problem.** To resolve it you need to find every seed with a config setting for `my_column_b` _and then cross-reference it with the seeds themselves_ 🫠 

Instead of letting `agate` raise the warning, we should first check the columns we are handing to agate, and the column names we're handing to agate to force the type of. If a column name with a type specification doesn't exist in the columns of the seed, we (`dbt-core`) should raise a warning with better information about the seed, and then also not hand that column type specification to agate.

### Acceptance criteria

* we raise a warning with seed information when a column type is specified for a column that doesn't exist in a seed
* agate doesn't raise a warning about the same column

### Suggested Tests

1. Test that we get our warning
2. Test that we don't get the agate warning

### Impact to Other Teams

NA

### Will backports be required?

No

### Context

_No response_

### PR DESCRIPTION
**Pull request by abhishek09827**
Resolves #12379

### Problem

Currently, [dbt](cci:1://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/core/dbt/context/providers.py:1133:4-1135:34) does not explicitly validate if the columns specified in `config.column_types` for a seed actually exist in the CSV file. If a user defines a type for a non-existent column, it might cause confusion or be silently ignored by the underlying library without clear feedback. This PR aims to improve the developer experience by validating these inputs and providing helpful warnings.

### Solution

This PR implements validation logic to ensuring that [column_types](cci:1://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:128:4-146:37) specified in the seed configuration are present in the actual CSV file.

**Implementation Details:**

*   **Fixed Column Type Validation Logic ([dbt-core/core/dbt/context/providers.py](cci:7://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/core/dbt/context/providers.py:0:0-0:0))**:
    *   Modified the [load_agate_table()](cci:1://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/core/dbt/context/providers.py:1253:4-1308:20) method to validate column existence before passing [column_types](cci:1://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:128:4-146:37) to agate.
    *   Reads the CSV file header to identify available columns.
    *   Checks if each column defined in [column_types](cci:1://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:128:4-146:37) config exists in the header.
    *   Emits a [dbt-core](cci:7://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core:0:0-0:0) `JinjaLogWarning` for any invalid columns with clear context.
    *   Removes invalid columns from the [column_types](cci:1://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:128:4-146:37) dictionary passed to agate to prevent downstream issues.
    *   Handles edge cases such as file read errors, empty files, or invalid delimiters.

*   **Warning Message Format**:
    When a column type is specified for a non-existent column, [dbt-core](cci:7://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core:0:0-0:0) now emits:
    > Column type specified for non-existent column '{column_name}' in seed '{seed_name}' (file: {path}). This column type override will be ignored.

*   **Comprehensive Tests ([dbt-core/tests/functional/seeds/test_seed_column_type_validation.py](cci:7://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:0:0-0:0))**:
    *   [TestSeedColumnTypeValidation](cci:2://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:60:0-110:43): Tests seeds with invalid column types to ensure warnings are raised un-needed types are ignored.
    *   [TestSeedColumnTypesBasic](cci:2://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:113:0-146:37): Tests basic seed column types functionality to ensure no regressions.
    *   [TestSeedColumnTypesWithMismatchedColumns](cci:2://file:///e:/Personal%20Projects/OpenSource/dbt/dbt-core/tests/functional/seeds/test_seed_column_type_validation.py:149:0-201:59): Tests seeds where column types reference non-existent CSV columns.

**Verification:**
All acceptance criteria have been met:
*   [x] Validates column existence before passing to agate
*   [x] Emits dbt-core warning with clear context (seed name, file path, invalid column name)
*   [x] Excludes invalid columns from column types passed to agate
*   [x] Agate does not emit its own warning for this case
*   [x] Behavior is non-fatal (warning only)

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

File: .changes/unreleased/Fixes-20260125-113244.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+kind: Fixes
+body: '[Bug] Unhelpful bug from agate when column type specified for non-existent column'
```
Comment (ash2shukla) [inline .changes/unreleased/Fixes-20260125-113244.yaml]: body should be telling what u fixed eg. `Only pass existing seed columns to agate and fire warning for non-existing columns specified in column types`

Comment (ash2shukla) [inline .changes/unreleased/Fixes-20260125-113244.yaml]: Issue: 12379 ( without # )
body: <should be what's changing not the issue itself imo> eg. Validate and filter column type for seeds

File: core/dbt/context/providers.py [hunk 2]
```diff
@@ -1269,8 +1271,37 @@ def load_agate_table(self) -> "agate.Table":
 
         column_types = self.model.config.column_types
         delimiter = self.model.config.delimiter
+
+        # Validate that column_types keys exist in the file header
+        filtered_column_types = column_types
+        if column_types:
+            try:
+                # Use utf-8-sig to handle BOM if present
+                with open(path, "r", encoding="utf-8-sig", newline="") as f:
```
Comment (ash2shukla) [inline core/dbt/context/providers.py]: Can be prettified a little -
```python
def _get_csv_header(path) -> Set[str]:
    with open(....):
       return ...
```

```python

column_types = self.model.config.column_types or {}

csv_header = _get_csv_header(path)
valid_column_types = set(column_types) & csv_header

invalid_column_names = set(column_types) - csv_header
if invalid_column_names:
   invalid_column_names_str = ', '.join(invalid_column_names)
   msg = f"These columns specified in column types do not exist in seed {self.model.name} ( file: {path}). These will be ignored: {invalid_column_names_str}"
   fire_event(SomeWarning)
```

File: tests/functional/seeds/test_seed_column_type_validation.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        # Find the other_seed result
+        other_seed_result = next((r for r in results if r.node.name == "other_seed"), None)
+        assert other_seed_result is not None
+
+        # Check that the agate table was created
+        assert other_seed_result.agate_table is not None
+
+        # Verify the table has the correct columns
+        column_names = [col.name for col in other_seed_result.agate_table.columns]
+        print(f"DEBUG: Columns in 'other_seed' table: {column_names}")
```
Comment (ash2shukla) [inline tests/functional/seeds/test_seed_column_type_validation.py]: Can u remove these DEBUG: statements and any comments that are unncessary eg L166

**Review by ash2shukla — CHANGES_REQUESTED** [review]
Thanks for your PR ! Can u please address these comments ?

**Comment by abhishek09827** [comment]
> Thanks for your PR ! Can u please address these comments ?

Hi @ash2shukla, Thanks for the review I've addressed your comments with the latest commit.

File: core/dbt/context/providers.py [hunk 4]
```diff
@@ -1249,6 +1252,17 @@ def try_or_compiler_error(
         except Exception:
             raise CompilationError(message_if_exception, self.model)
 
+    @staticmethod
+    def _read_csv_header(path: str, delimiter: str) -> Optional[Set[str]]:
```
Comment (MichelleArk) [inline core/dbt/context/providers.py:1256]: Conceptually it doesn't feel like `_read_csv_header` needs to belong on `ProviderContext`, especially since it doesn't require any state of the class. To avoid this class getting too bloated, perhaps we could add this method as public in a new module `core/dbt/clients/csv_helper.py` and unit test the method in isolation.

**Comment by MichelleArk** [comment]
Merging in main should address the integration test failures, fix was merged earlier this week here: https://github.com/dbt-labs/dbt-core/pull/12506

File: core/dbt/context/providers.py (hunk 4 above)
Comment (abhishek09827) [reply core/dbt/context/providers.py:1256]: Hi @MichelleArk,
Thanks for the feedback! You're right that `_read_csv_header` doesn't require instance state and the static method approach is cleaner.
Since this is a non-blocking issue, I plan to address it in a follow-up PR:
1. Create `core/dbt/clients/csv_helper.py` with a public `read_csv_header()` function
2. Move the `_read_csv_header` logic there
3. Add unit tests in `tests/unit/core/dbt/clients/test_csv_helper.py`
4. Update `ProviderContext` to use the new helper

Let me know if you'd prefer me to include this refactoring in the current PR instead

**Comment by abhishek09827** [comment]
Hi @MichelleArk ,

Just a gentle follow-up on this PR. I’ve addressed the previous feedback and made the requested changes.
Whenever you get a chance, could you please take another look and let me know if any further changes are needed?

Thanks!

**Comment by abhishek09827** [comment]
Thanks for the reviews and approvals! 
I’ve re-triggered CI after the earlier failing check. It now looks like the remaining required workflows are awaiting maintainer approval to run.

Could someone please approve the pending workflows when convenient? Happy to provide any additional clarification if needed. Thank you!

**Comment by abhishek09827** [comment]
Hi @MichelleArk,
I'm seeing CI job failures related to missing database roles and columns, e.g.:

```
FATAL: database "root" does not exist
FATAL: role "root" does not exist
ERROR: column "default_value", "meta_default_value", or "my_meta_value" does not exist
```
I wanted to confirm:
Does the CI environment provision these roles, databases, and columns when running tests, or should this be handled in the PR's test setup/scripts?
Please advise if there's anything I should add or update to resolve these errors and ensure the tests run successfully!

Thank you!

**Comment by MichelleArk** [comment]
Merging in main should address the failing test `TestWarnErrorOptionsFromProjectCanExcludeSpecificEvent`

**Comment by abhishek09827** [comment]
> Merging in main should address the failing test `TestWarnErrorOptionsFromProjectCanExcludeSpecificEvent`

Thanks for clarifying the reason for the failing test.
I’m currently unable to merge I am getting,
```
Merging is blocked
You're not authorized to push to this branch. 
```
Could you please advise on the next steps from here?

**Comment by MichelleArk** [comment]
Is this the error you're getting trying to update your branch with the commits in main?

**Comment by abhishek09827** [comment]
Thank you for your help, there was some issue from my end.
I have rebased the branch to upstream/main, please approve the pending workflows when convenient
